zkevm-circuits/src/evm_circuit/op_execution/byte.rs

```diff
@@ -0,0 +1,350 @@
+        will_halt: false,
+    };
+
+    pub(crate) fn construct(alloc: &mut CaseAllocation<F>) -> Self {
+        Self {
+            case_selector: alloc.selector.clone(),
+            index: alloc.words.pop().unwrap(),
+            value: alloc.words.pop().unwrap(),
+            result: alloc.words.pop().unwrap(),
+            is_msb_sum_zero: IsZeroGadget::construct(alloc),
+            is_byte_selected: array_init(|_| IsEqualGadget::construct(alloc)),
+        }
+    }
+
+    pub(crate) fn constraint(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+        name: &'static str,
+    ) -> Constraint<F> {
+        let mut cb = ConstraintBuilder::default();
+
+        // If any of the non-LSB bytes of the index word are non-zero we never need to copy any bytes.
+        // So just sum all the non-LSB byte values here and then check if it's non-zero
+        // so we can use that as an additional condition when to copy the byte value.
+        let msb_sum_zero = self
+            .is_msb_sum_zero
+            .constraints(&mut cb, utils::sum::expr(&self.index.cells[1..32]));
+
+        // Now we just need to check that `result[0]` is the sum of all copied bytes.
+        // We go byte by byte and check if `idx == index[0]`.
+        // If they are equal (at most once) we add the byte value to the sum, else we add 0.
+        // The additional condition for this is that none of the non-LSB bytes are non-zero (see above).
+        // At the end this sum needs to equal `result[0]`.
+        let mut select_sum = 0.expr();
+        for idx in 0..32 {
+            // Check if this byte is selected looking only at the LSB of the index word
+            let byte_selected = self.is_byte_selected[idx].constraints(
+                &mut cb,
+                self.index.cells[0].expr(),
+                (31 - idx).expr(),
+            );
+
+            // Add the byte to the sum when this byte is selected
+            select_sum = select_sum
+                + (byte_selected
+                    * msb_sum_zero.clone()
+                    * self.value.cells[idx].expr());
+        }
+        cb.require_equal(self.result.cells[0].expr(), select_sum);
+
+        // All bytes of result, except for the LSB, always need to be 0.
+        for idx in 1..32 {
+            cb.require_zero(self.result.cells[idx].expr());
+        }
+
+        // Pop the byte index and the value from the stack, push the result on the stack
+        cb.stack_pop(self.index.expr());
+        cb.stack_pop(self.value.expr());
+        cb.stack_push(self.result.expr());
+
+        // State transitions
+        #[allow(clippy::needless_update)]
+        utils::StateTransitions {
+            gc_delta: Some(GC_DELTA.expr()),
+            sp_delta: Some(SP_DELTA.expr()),
+            pc_delta: Some(PC_DELTA.expr()),
+            gas_delta: Some(GAS.expr()),
+            ..Default::default()
```

Comment:
nitpick: My clippy seems happy with the suggested update.
```suggestion
        utils::StateTransitions {
            gc_delta: Some(GC_DELTA.expr()),
            sp_delta: Some(SP_DELTA.expr()),
            pc_delta: Some(PC_DELTA.expr()),
            gas_delta: Some(GAS.expr()),
```